ckanext/dalrrd_emc_dcpr/plugin.py

```diff
@@ -15,3 +21,83 @@ def update_config(self, config_):
 
     def get_commands(self):
         return [test_ckan_cmd]
+
+    # IAuthFunctions
+
+    def get_auth_functions(self) -> typing.Dict[str, typing.Callable]:
+        return {
+            "package_publish": authorize_package_publish,
+        }
+
+    def get_actions(self) -> typing.Dict[str, typing.Callable]:
+        return {
+            "package_create": package_create,
+            "package_update": package_update,
+            "package_patch": package_patch,
+        }
+
+
+def authorize_package_publish(
+    context: typing.Dict, data_dict: typing.Optional[typing.Dict] = None
+) -> typing.Dict[str, bool]:
+    # in here we can inspect the context and the data_dict in order to
+    # decide whether to authorize access or not
+    user_name = context.get("user")
+    owner_org = data_dict.get("owner_org") or data_dict.get("group_id")
+
+    members = toolkit.get_action("member_list")(
+        data_dict={"id": owner_org, "object_type": "user"}
+    )
+    admin_member_ids = [
+        member_tuple[0] for member_tuple in members if member_tuple[2] == "Admin"
+    ]
+    convert_user_name_or_id_to_id = toolkit.get_converter(
+        "convert_user_name_or_id_to_id"
+    )
+    user_id = convert_user_name_or_id_to_id(user_name, context)
+
+    if user_id in admin_member_ids:
+        return {"success": True}
+    else:
+        logger.debug(f"Inside authorize_package_publish function: {locals()}")
+        return {"success": False, "msg": "You are not authorized to publish a package"}
+
+
+# IActions
+
+
+@toolkit.chained_action
+def package_create(original_action, context, data_dict):
+    """
+    Intercepts the core `package_create` action to check if package
+     is being published after being created.
+    """
+    return package_publish_check(original_action, context, data_dict)
+
+
+@toolkit.chained_action
+def package_update(original_action, context, data_dict):
+    """
+    Intercepts the core `package_update` action to check if package is being published.
+    """
+    return package_publish_check(original_action, context, data_dict)
+
+
+@toolkit.chained_action
+def package_patch(original_action, context, data_dict):
+    """
+    Intercepts the core `package_patch` action to check if package is being published.
+    """
+    return package_publish_check(original_action, context, data_dict)
+
+
+def package_publish_check(action, context, data):
+    remains_private = toolkit.asbool(data.get("private", True))
+    result = None
+    if remains_private:
+        # make sure private state is updated, then assign result
+        data.update({"private": remains_private})
```

Comment:
This is not needed. We are not modifying the `data` dict in any way prior to this, so no need to update it.

Reply:
There are cases that we will need to do this.

For now we are hiding the visibility input for users with role `editor` that means when the corresponding users are creating datasets the `data` dict wont have a `private` property, the CKAN default behavior is if the `data` doesn't have the `private` property it automatically save the dataset and set its `private` property to `False`.
This can cause problems to user when adding datasets resources because the `package_publish` check will be evaluated when they add resources and now the dataset will be public, so the will be denied to add a resource to a dataset that they still creating.

Reply:
I don't like the fact that we are modifying the `data_dict` in a function that is supposed to perform package creation.

What we are doing by hiding the `private` form field when the user is not authorized to modify it is merely a User Experience refinement and ideally it should not impact the backend logic. 

What about implementing this at the template level?

Lets code the template in such a way as to always send in the `private` field, regardless of it being show to the user or not. If the form field is visible, then the user will fill it with the appropriate value. If the form field is not to be shown to the user we can use something like `<input type="hidden" name="private" value="True">`. This means we'd always be sending back a `private` field to the backend that processes the form.

Reply:
Right, that will do. What about the CKAN API, how do you suggest we go about that? At the moment the API allows sending the data without the `private` property and in turn CKAN assign the corresponding dataset `private` property a `False` value.